src/Graph.ts

```diff
@@ -121,44 +118,15 @@ export default class Graph {
 			return this.acornParse(code, { ...defaultAcornOptions, ...options, ...this.acornOptions });
 		};
 
-		const assetPluginHooks = createAssetPluginHooks(this.assetsById);
-
-		this.pluginContext = {
-			watcher,
-			isExternal: undefined,
-			resolveId: undefined,
-			parse: this.contextParse,
-			warn: (warning: RollupWarning | string) => {
-				if (typeof warning === 'string') warning = { message: warning };
-				this.warn(warning);
-			},
-			error: (err: RollupError | string) => {
-				if (typeof err === 'string') throw new Error(err);
-				error(err);
-			},
-			...assetPluginHooks
-		};
+		this.pluginDriver = createPluginDriver(this, options, watcher);
 
-		this.resolveId = first(
-			[
-				((id: string, parentId: string) =>
-					this.isExternal(id, parentId, false) ? false : null) as ResolveIdHook
-			]
-				.concat(
-					this.plugins
-						.map(plugin => plugin.resolveId)
-						.filter(Boolean)
-						.map(resolveId => resolveId.bind(this.pluginContext))
-				)
-				.concat(resolveId(options))
-		);
-
-		this.pluginContext.resolveId = this.resolveId;
-
-		const loaders = this.plugins.map(plugin => plugin.load).filter(Boolean);
-		this.hasLoaders = loaders.length !== 0;
-
-		this.load = first([...loaders, load]);
+		if (typeof options.external === 'function') {
+			this.isExternal = options.external;
+		} else {
+			const external = options.external;
+			const ids = new Set(Array.isArray(external) ? external : external ? [external] : []);
+			this.isExternal = id => ids.has(id);
```

Comment:
Definitely an improvement! As all the ids are strings, we could also put them as keys into an object `{[id: string]: true}` for another very slight performance boost? Might be a micro-optimization, though.

Reply:
JS sets should be equal or better in performance according to theory.

Reply:
Definitely looks a little cleaner so let's follow the theory. Googling shows that results are definitely inconclusive which is not necessarily what I would have expected, so the cleaner solution should win here 👍